File: commercehub-identity/src/main/java/com/commercehub/identity/service/UserService.java
package com.commercehub.identity.service;

import com.commercehub.identity.dto.AddressResponse;
import com.commercehub.identity.dto.CreateAddressRequest;
import com.commercehub.identity.dto.UpdateAddressRequest;
import com.commercehub.identity.dto.AssignRoleRequest;
import com.commercehub.identity.dto.ChangePasswordRequest;
import com.commercehub.identity.dto.UpdateUserRequest;
import com.commercehub.identity.dto.UserResponse;
import com.commercehub.identity.dto.UserSummaryResponse;
import com.commercehub.identity.entity.UserStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface UserService {

    UserResponse getOwnProfile(UUID userId);

    UserResponse updateOwnProfile(UUID userId, UpdateUserRequest request);

    Page<UserSummaryResponse> getAllUsers(Pageable pageable);

    UserResponse getUserById(UUID id);

    UserResponse updateUserStatus(UUID id, UserStatus status);

    List<AddressResponse> getAddresses(UUID userId);

    AddressResponse addAddress(UUID userId, CreateAddressRequest request);

    AddressResponse updateAddress(UUID userId, UUID addressId, UpdateAddressRequest request);

    void deleteAddress(UUID userId, UUID addressId);

    void setDefaultAddress(UUID userId, UUID addressId);

    void changePassword(UUID userId, ChangePasswordRequest request);

    UserResponse assignRole(UUID userId, AssignRoleRequest request);

    UserResponse removeRole(UUID userId, AssignRoleRequest request);
}

File: commercehub-frontend/src/features/authentication/services/userService.ts
import { apiClient } from '@/api/client';
import type { ApiResponse, PaginatedResponse } from '@/types/api';
import type {
  AddressResponse,
  CreateAddressRequest,
  UpdateAddressRequest,
  UpdateUserRequest,
  UpdateUserStatusRequest,
  UserResponse,
  UserSummaryResponse,
} from '../types';

export const userService = {
  getMe: async (accessToken?: string): Promise<UserResponse> => {
    const config = accessToken ? { headers: { Authorization: `Bearer ${accessToken}` } } : {};
    const response = await apiClient.get<ApiResponse<UserResponse>>('/users/me', config);
    return response.data.data;
  },

  changePassword: async (data: { currentPassword: string; newPassword: string }): Promise<void> => {
    await apiClient.post('/users/me/change-password', data);
  },

  assignRole: async (id: string, role: string): Promise<UserResponse> => {
    const response = await apiClient.post<ApiResponse<UserResponse>>(`/users/${id}/roles`, { role });
    return response.data.data;
  },

  removeRole: async (id: string, role: string): Promise<UserResponse> => {
    const response = await apiClient.delete<ApiResponse<UserResponse>>(`/users/${id}/roles`, { data: { role } });
    return response.data.data;
  },

  updateMe: async (data: UpdateUserRequest): Promise<UserResponse> => {
    const response = await apiClient.patch<ApiResponse<UserResponse>>('/users/me', data);
    return response.data.data;
  },

  getUserById: async (id: string): Promise<UserResponse> => {
    const response = await apiClient.get<ApiResponse<UserResponse>>(`/users/${id}`);
    return response.data.data;
  },

  listUsers: async (params: {
    page?: number;
    size?: number;
  }): Promise<PaginatedResponse<UserSummaryResponse>> => {
    const response = await apiClient.get<PaginatedResponse<UserSummaryResponse>>('/users', {
      params,
    });
    return response.data;
  },

  updateUserStatus: async (id: string, data: UpdateUserStatusRequest): Promise<UserResponse> => {
    const response = await apiClient.patch<ApiResponse<UserResponse>>(
      `/users/${id}/status`,
      data,
    );
    return response.data.data;
  },

  getAddresses: async (): Promise<AddressResponse[]> => {
    const response = await apiClient.get<ApiResponse<AddressResponse[]>>('/users/me/addresses');
    return response.data.data;
  },

  addAddress: async (data: CreateAddressRequest): Promise<AddressResponse> => {
    const response = await apiClient.post<ApiResponse<AddressResponse>>(
      '/users/me/addresses',
      data,
    );
    return response.data.data;
  },

  updateAddress: async (id: string, data: UpdateAddressRequest): Promise<AddressResponse> => {
    const response = await apiClient.put<ApiResponse<AddressResponse>>(
      `/users/me/addresses/${id}`,
      data,
    );
    return response.data.data;
  },

  deleteAddress: async (id: string): Promise<void> => {
    await apiClient.delete(`/users/me/addresses/${id}`);
  },
};
